src/platform/graphics/shader-utils.js

```diff
@@ -50,22 +50,35 @@ class ShaderUtils {
      * @param {string} [options.fragmentDefines] - The fragment shader defines.
      * @param {string} [options.fragmentExtensions] - The fragment shader extensions code.
      * @param {string} [options.fragmentPreamble] - The preamble string for the fragment shader.
-     * @param {boolean} [options.useTransformFeedback] - Whether to use transform feedback. Defaults
-     * to false.
+     * @param {boolean} [options.useTransformFeedback] - Whether to use transform feedback. Defaults to false.
+     * @param {string | string[]} [options.fragmentOutputTypes] - Fragment shader output types, which default to vec4. Passing a string will set the output type for all color attachments. Passing an array will set the output type for each color attachment.
      * @returns {object} Returns the created shader definition.
      */
     static createDefinition(device, options) {
         Debug.assert(options);
 
-        const getDefines = (gpu, gl2, gl1, isVertex) => {
+        const getDefines = (gpu, gl2, gl1, isVertex, options) => {
 
             const deviceIntro = device.isWebGPU ? gpu :
                 (device.isWebGL2 ? gl2 : ShaderUtils.gl1Extensions(device, options) + gl1);
 
             // a define per supported color attachment, which strips out unsupported output definitions in the deviceIntro
             let attachmentsDefine = '';
-            for (let i = 0; i < device.maxColorAttachments; i++) {
-                attachmentsDefine += `#define COLOR_ATTACHMENT_${i}\n`;
+
+            // Normalize fragmentOutputTypes to an array
+            let fragmentOutputTypes = options.fragmentOutputTypes ?? ['vec4'];
+            if (!Array.isArray(fragmentOutputTypes)) {
+                fragmentOutputTypes = [fragmentOutputTypes ?? 'vec4'];
+            }
+
```

Comment:
slightly simplified.
```
            let fragmentOutputTypes = options.fragmentOutputTypes ?? 'vec4';
            if (!Array.isArray(fragmentOutputTypes)) {
                fragmentOutputTypes = [fragmentOutputTypes];
            }
```
But also, this can be inside the `if (!isVertex)` condition, it does not need to run otherwise.